FAQ.md: Add reformatted Background.md contents (#41)
TrenchBoot/TrenchBoot.github.io: docs/FAQ.md

# TrenchBoot FAQ

1. [Why use Trenchboot?](#1-why-use-trenchboot-background-info)
2. [How does TrenchBoot work?](
    #2-how-does-trenchboot-work-trenchboot-architecture)
3. [How to use Trenchboot?](#3-how-to-use-trenchboot-practical-usage)
4. [Why does TrenchBoot use an intermediate launcher?](
    #4-why-does-trenchboot-use-an-intermediate-launcher)
5. [What are the benefits of measurement over signature validation?](
    #5-what-are-the-benefits-of-measurement-over-signature-validation)
6. [What do I need to incorporate TrenchBoot into my system?](
    #6-what-do-i-need-to-incorporate-trenchboot-into-my-system)
7. [Where do I start if I want to help with contributions?](
    #7-where-do-i-start-if-i-want-to-help-with-contributions)
8. [Where to find the sources?](#8-trenchboot-sources)
9. [Various GitHub and further documentation links](
    #9-various-github-and-further-documentation-links)

> ## 1. Why use Trenchboot? Background info

Reviewer: Maybe add the second part in parenthesis?  As is it looks as an answer to the question, which it isn't.
```suggestion
## 1. Why use Trenchboot? (background info)
```
The same applies to questions 2 and 3.

Author: That's a good point. https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/021546389649000c5c047f9885295264f99b51bf

### What is Trust?

> Trust is when the computer does something on our behalf we have to trust it

Reviewer: > on our behalf we have to trust it

This looks wrong, like a preposition or punctuation is missing.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/333691d2321d4a1a49e964dd41f811f4dda7daa5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R23

with our data. A trusted component is that which always behaves in an expected
manner.  To make a component "trusted", you must establish some root of trust.
A Root of Trust Measurement is the process of measuring (hashing) components on
a system to validate their integrity. When establishing a root of trust, there
are two primary schools of thought: Static Root of Trust Measurements and
Dynamic Root of Trust Measurements.

### SRTM

SRTM takes place at system boot from firmware / bios. This process requires
measurements of the entire boot sequence at once. This is both a good and a
bad thing, because any changes to any part of the process (including even
bios settings) would cause the entire system to need to be remeasured. SRTM

> Starts with Core Root of Trust (CRTM), and works its way up the chain where

Reviewer: ```suggestion
Starts with Core Root of Trust for Measurement (CRTM), and works its way up the chain where
```

each link verifies the next. In other words, SRTM starts by measuring the bios
boot block and then moves on to measure the rest of the bios.

### DRTM

DRTM aims to address shortcomings of SRTM. DRTM creates trusted environment from
an untrusted state. Essentially what this means is that regardless of weather
or not computer was booted from trusted software, it uses hardware to ensure
that no malicious software has been run. For this reason, DRTM takes place
after boot, as opposed to SRTM, which takes place pre-boot. Intel's solution
for DRTM is TXT, which utilizes special instructions on the processor (SMX).
AMD's solution for DRTM is called SVM. Trenchboot provides a solution for

> setting up DRTM for an arbitrary platform

Reviewer: We're saying AMD SKINIT in most places.  Also, I think SKINIT depends on SVM, but virtualization can do just fine without SKINIT.
```suggestion
AMD's solution for DRTM is called SKINIT. Trenchboot provides a solution for
setting up DRTM for an arbitrary platform.
```

### Typical Measured Launch Process

The process for carrying out DTRM depends on your system.
[Intel TXT Spec](
    https://software.intel.com/content/www/us/en/develop/articles/intel-trusted-execution-technology-intel-txt-enabling-guide.html)

> [AMD SVM Spec](https://www.amd.com/system/files/TechDocs/24593.pdf)

Reviewer: To fix formatting (currently the links appear after the sentence):
```suggestion
- [Intel TXT Spec](
    https://software.intel.com/content/www/us/en/develop/articles/intel-trusted-execution-technology-intel-txt-enabling-guide.html)
- [AMD SVM Spec](https://www.amd.com/system/files/TechDocs/24593.pdf)
```

### Why TrenchBoot?

- Managing each of these components individually can be confusing
- Most DTRM systems are platform dependent
- Most DTRM systems can only measure what was loaded into memory with it at the
    time it was started

> - Most DTRM systems support only one attestation action

Reviewer: Essentially repeating the question looks confusing.  The first bullet confusingly refers to "these components", typos and more confusing third bullet.

```suggestion
### So, why TrenchBoot?

- Managing multiple independent components individually can be confusing
- Most DRTM systems are platform-dependent
- Most DRTM systems can only measure entities already present in RAM at the
    time DRTM sequence is initiated
- Most DRTM systems support only one attestation action
```

## 2. How does TrenchBoot work? Trenchboot Architecture

### Architecture Overview

*The full architecture overview can be found at

> [trenchboot.org](https://trenchboot.org/documentation/Architecture)*

Reviewer: ```suggestion
on [this page](/documentation/Architecture)*
```

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/a6c1875dc0521a25062069627278bdcc3727a7b7

Reviewer: This wasn't addressed, the commit corresponds to a different comment.

Author: Sorry, my mistake https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/bd655a0cc26a66c1f7143ed1a4bd48c026ee39e6

![Overview](img/Architectural_Flow.png)
This diagram shows a general overview of the overall architecture of a
TrenchBoot system. In Bootstrap phase the system environment is measured and
that information stored in the dynamic PCRs, using whatever method is
appropriate for the given hardware. The intermediate phase analyzes the
information collected in the bootstrap phase to decide if it wants to continue
booting and pass control to the target OS, or if it wants to stop immediately.
If desired, additional evidence/measurements can be gathered in this phase. Once
 the integrity of the system is verified the TrenchBoot kernel passes off
 control to the actual desired OS to initiate the runtime phase.

### Bootstrap Phase - GRUB

Typically, when attempting to establish a DRTM, TBoot is loaded in order to

> carry out TXT on Intel, or SEV is used on AMD. TrenchBoot instead utilizes its

Reviewer: ```suggestion
carry out TXT on Intel, or SKINIT is used on AMD. TrenchBoot instead utilizes its
```

Author: commited using gh UI

Reviewer: Doesn't look like it.

Author: Yes, I must have did something wrong. https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/5fede969ab731b1b1a5509a5d95c77a6aac861b9

own system, which is a custom version of GRUB that has commands built in to

> carry out a TrenchBoot "Secure Launch". This customized grub introduces the

Reviewer: ```suggestion
carry out a TrenchBoot "Secure Launch". This customized GRUB introduces the
```

Author: 6ff4a2f

> slaunch and slaunch_module commands. A working example of the TrenchBoot GRUB

Reviewer: ```suggestion
`slaunch` and `slaunch_module` commands. A working example of the TrenchBoot GRUB
```

Author: f7b7a6c

can be found at
[https://github.com/3mdeb/meta-trenchboot](
    https://github.com/3mdeb/meta-trenchboot).
The custom GRUB is however unfinished and while it is running on AMD, it is not

> yet functional with Intel processors.

Reviewer: This information is outdated.
```suggestion
```
The current state is that it works for Linux and Multiboot2 protocols for both Intel and AMD, but only in case of legacy boot.  Maybe that should be added instead.

Author: [2c5547e](https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/2c5547e1a2457fb56055d3dbeb8620433b405f41)

### Intermediate Phase - Linux Kernel / U-ROOT (Initramfs)

The intermediate phase processes information collected in the bootstrap phase
and converts it into normalized data. This data is eventually used to determine
whether to boot into the target OS. The kernel and u-root initramfs
together make up Trenchboot's "Security Engine", which is an intermediate
mini-OS that serves to process data gathered by that bootstrap phase. During the
intermediate phase, the kernel and initramfs work together to measure block
devices, individual files, SMBUS/DMI information and provides unseal based,
external device and network based attestation. Trenchboot calls this
functionality secure launch (aka slaunch). The table in section 4 lists the

> specific repo and branch to use when building the kernel or u-root.

Reviewer: This section looks outdated or at least missing an implementation if there are still plans for it.  ACM/secure-kernel-loader look more like an intermediate phase.

Author: I don't know about that. Someone else should decide on this section then.

Reviewer 2: Possibly related to first diagram on https://trenchboot.org/documentation/Architecture/#general-architecture and https://trenchboot.org/documentation/DevelopersGuide/#trenchboot-linuxu-root-configuration. The idea was that there would be a minimal Linux that is launched dynamically, and it then would kexec to the target OS.

This section has some dubious parts. Measuring block devices would take forever, SMBUS probably was supposed to be SMBIOS, and intermediate phase doesn't provide network based attestation. I have no idea to what `table in section 4` refers to.

## 3. How to use Trenchboot? Practical usage

### Trenchboot GRUB TXT Config with Serial Communication

*Note: ACM modules can be downloaded from

> [mirror.openxt.org](http://mirror.openxt.org/)

Reviewer: The site looks unavailable.  Maybe point to https://www.intel.com/content/www/us/en/developer/articles/tool/intel-trusted-execution-technology.html

Author: Yes, I forgot to mention it. I was unsure if this was a temporary problem because my browser seemed to try to load it indefinitely. If you a know a better site to link to instead then that's great.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/compare/77aaa7d4fca79f07460ba2bc9372c2e0cea0e01b..333691d2321d4a1a49e964dd41f811f4dda7daa5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R115

```text
    menuentry 'TrenchBoot' {
            savedefault
            load_video
            set gfxpayload=keep
            insmod gzio
            insmod part_msdos
            insmod ext2
            set root='hd0,msdos1'
            multiboot2 /boot/slboot.gz loglvl=all serial=115200,8n1,0x3f8 logging=serial,memory
            module2 /boot/vmlinuz-trenchboot root=UUID=<your block uuid> ro intel_iommu=on loglevel=7 console=ttyS0,115200 nokaslr 
            module2 /boot/<your acm>.bin

>             module2 /boot/initramfs-trenchboot.cpio

Reviewer: This should be replaced by
```
slaunch
slaunch_module /boot/<your acm>.bin
```
It's usually done above specifying kernel/initrd, but might work after as well.

And again, `slboot.gz`, `/boot/vmlinuz-trenchboot` might not be relevant at this point.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/compare/77aaa7d4fca79f07460ba2bc9372c2e0cea0e01b..333691d2321d4a1a49e964dd41f811f4dda7daa5?diff=split&w=#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17L126-R128

Reviewer: Oups, I didn't mean to suggest just dropping slboot/vmlinuz lines.  They either need to be kept or replaced with something else.  Please bring them back, as an example it probably doesn't matter much if this is oudated.

Author: Sure, sorry, I have no experience in configuring GRUB. https://github.com/TrenchBoot/TrenchBoot.github.io/compare/2c5547e1a2457fb56055d3dbeb8620433b405f41..a140413cd7375ed2e219946580c574a1d28ac7e5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R126-R127

    }
```

### Testing the System

QEMU has been tested to boot fully into u-root while simulating the Trenchboot
kernel.
QEMU For external u-root

```bash
qemu-systemx86_64 -kernel <kernel repo>/arch/x86/boot/bzImage -initrd /tmp/initramfs.linux_amd64.cpio -append console=ttyS0,38400 -serial file:serial.out
```

QEMU with u-root built-in

```bash
qemu-systemx86_64 -kernel <kernel location>
```

> ## 4. Why does TrenchBoot use an intermediate launcher?

Reviewer: It doesn't.  Was it completely abandoned?  Is it planned to be reintroduced again?  

It's not part of these changes, but answers to these questions define what we should keep or import.  The section above about running Linux with u-root in QEMU isn't relevant without the intermediate launcher.

@krystian-hebel, @andyhhp, @rossphilipson.

Reviewer 2: I think @dpsmith could elaborate on this, as well as https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41#discussion_r1739735720.

# TrenchBoot FAQ

1. [Why use Trenchboot?](#1-why-use-trenchboot-background-info)
2. [How does TrenchBoot work?](
    #2-how-does-trenchboot-work-trenchboot-architecture)
3. [How to use Trenchboot?](#3-how-to-use-trenchboot-practical-usage)
4. [Why does TrenchBoot use an intermediate launcher?](
    #4-why-does-trenchboot-use-an-intermediate-launcher)
5. [What are the benefits of measurement over signature validation?](
    #5-what-are-the-benefits-of-measurement-over-signature-validation)
6. [What do I need to incorporate TrenchBoot into my system?](
    #6-what-do-i-need-to-incorporate-trenchboot-into-my-system)
7. [Where do I start if I want to help with contributions?](
    #7-where-do-i-start-if-i-want-to-help-with-contributions)
8. [Where to find the sources?](#8-trenchboot-sources)
9. [Various GitHub and further documentation links](

>     #9-various-github-and-further-documentation-links)

Reviewer 2: I think those should be removed. Original PR predates current form of website, and ideally the website should be easy enough to navigate so that questions like these wouldn't be asked in the first place.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/498ff32bb978a87aa3a6c1af0c20ed1049515a27

> ## 1. Why use Trenchboot? (background info)

Reviewer 2: `TrenchBoot`, here and everywhere in this file.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/671b0dd276e69a1846fe77e6de708fb34284a3fb

### What is Trust?

Trust is when the computer does something on our behalf and we have to trust it
with our data. A trusted component is that which always behaves in an expected
manner.  To make a component "trusted", you must establish some root of trust.
A Root of Trust Measurement is the process of measuring (hashing) components on

> a system to validate their integrity. When establishing a root of trust, there

Reviewer 2: https://github.com/TrenchBoot/TrenchBoot.github.io/blob/cleanup/docs/theory/Glossary.md#trust has different, incompatible definition. Also, there is no `Root of Trust Measurement`, but there is `Root of Trust for Measurements`. It is not a process of measuring, though.

Author: On one hand, an explanation more verbose than a dictionary-like definition is more friendly for someone who would actually ask the question in the section header. On the other hand, it's not good that we try to define a term in two seperate places, and they are entirely different. Such small things may leave a bad opinion on a project that is all about trust. 

And as you said, it does sound wrong to measure the Root of Trust. We can not do that, otherwise it would not be the Root. We can determine if we trust something else by forming a chain of transient trust, like in PGP or x509.

Author: @krystian-hebel What do you think about rewriting it like that? https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/cf12d2ab521f30c9f1e9c34ee4076e660d5588e0 Is it right?

Author: Also, I think if the glossary defines such basic concepts like trust, trustee, trustor, it could also use defining the basic terms related to cybersecurity like Integrity, Confidentiality, Authenticity, Non-Repudiation, and basic techniques related to it, like a Hash, a Signature, a Certificate.

Reviewer 2: Not really, this isn't PKI, it is more generalized concept. I like the beginning, but please drop everything from `To make a component "trusted"` to `is identical to the one our Root of Trust assured we can trust by signing it.`

Also, `Static Root of Trust Measurements` -> `Static Root of Trust for Measurement` (add `for`, make `Measurement` singular), same for DRTM.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/compare/0463f5e6c506ce0ba96b957b98d8768e5a12415d..25021ec141d2a88fe039a75c8743ba4dd733f2e3

are two primary schools of thought: Static Root of Trust Measurements and
Dynamic Root of Trust Measurements.

### SRTM

> SRTM takes place at system boot from firmware / bios. This process requires

Reviewer 2: `bios` -> `BIOS`, everywhere in this file.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/cc3fe3b6f35b738c967fc5dcdaf8288d682e9870

> measurements of the entire boot sequence at once. This is both a good and a

Reviewer 2: I don't agree, this forms a chain of trust (as explained few sentences further), and not everything is measured at once because of that.

bad thing, because any changes to any part of the process (including even

> bios settings) would cause the entire system to need to be remeasured. SRTM

Reviewer 2: Entire system is measured on each boot anyway, regardless of changes made to configuration. I think this part mistook measurements and sealing of data to PCR values, those are two separate features.

Reviewer 2: It would probably be best to remove this and previous sentence.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/0f09b97b01642ebbfa22afba55f8ce54938c2c49

> Starts with Core Root of Trust for Measurement (CRTM), and works its way up the chain where

Reviewer 2: Wrap this line.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/cc3fe3b6f35b738c967fc5dcdaf8288d682e9870

each link verifies the next. In other words, SRTM starts by measuring the bios
boot block and then moves on to measure the rest of the bios.

### DRTM

> DRTM aims to address shortcomings of SRTM. DRTM creates trusted environment from

Reviewer 2: No shortcomings were mentioned in SRTM section. The most important ones are big TCB (Trusted Computing Base), the need to reboot the platform in order to reassure the validity of chain of trust and the fact that SRTM provides only a load-time integrity assurance. Please add those.

Author: > The most important ones are big TCB (Trusted Computing Base)

Does this mean that it needs more hardware and software support? That it causes the firmware to increase in size, or that more components are needed to allow for this mode of establishing trust?

> only a load-time integrity assurance

Meaning it is only useful during the boot process and does not help afterwards?

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/6f09dc7f5be5684e51237e8916ffbcce9b29f6f4

I'd like to amend this commit later with more details

Reviewer 2: > Does this mean that it needs more hardware and software support? That it causes the firmware to increase in size, or that more components are needed to allow for this mode of establishing trust?

Not quite, it doesn't necessarily _need_ it (at least not for maintaining trusted state), but more software is part of the chain of trust. For example, firmware has all of the hardware initialization code, even if it isn't directly responsible for keeping the chain intact. More code = more bugs, more work required to assess its security. With DRTM, system firmware is (mostly) not part of TCB.

> Meaning it is only useful during the boot process and does not help afterwards?

Universal answer to all security-related questions: it depends on the threat model. It means that integrity was good at the boot time, but doesn't guarantee that the same can be told about current state. Whether it is good enough depends on the threat model.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/8826bfd8b8c4b2d56e57862557060826ee004cdd

> an untrusted state. Essentially what this means is that regardless of weather

Reviewer 2: ```suggestion
an untrusted state. Essentially what this means is that regardless of whether
```

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/6f09dc7f5be5684e51237e8916ffbcce9b29f6f4#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R45

or not computer was booted from trusted software, it uses hardware to ensure

> that no malicious software has been run. For this reason, DRTM takes place

Reviewer 2: ```suggestion
that no malicious software, which may or may not been run before dynamic launch, can impact software started after dynamic launch. For this reason, DRTM takes place
```

Plus wrap it.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/69c288499a806881c689308184b58c1f3b83c21b

after boot, as opposed to SRTM, which takes place pre-boot. Intel's solution
for DRTM is TXT, which utilizes special instructions on the processor (SMX).
AMD's solution for DRTM is called SKINIT. Trenchboot provides a solution for
setting up DRTM for an arbitrary platform.

### Typical Measured Launch Process

The process for carrying out DTRM depends on your system:

- [Intel TXT: Software Development Guide][txt-spec]
- [System Programming Volume of for AMD64 (PDF)][skinit-spec] (see
  "Secure Startup with SKINIT" section)

[txt-spec]: https://www.intel.com/content/www/us/en/content-details/315168/intel-trusted-execution-technology-intel-txt-software-development-guide.html
[skinit-spec]: https://www.amd.com/system/files/TechDocs/24593.pdf

### So, why TrenchBoot?

- Managing multiple independent components individually can be confusing
- Most DRTM systems are platform-dependent
- Most DRTM systems can only measure entities already present in RAM at the
    time DRTM sequence is initiated

> - Most DRTM systems support only one attestation action

Reviewer 2: None of those answer the question put in the header. Please add a sentence or two how TrenchBoot tries to help with those issues, e.g. by paraphrasing first sentence of https://github.com/TrenchBoot/TrenchBoot.github.io/blob/cleanup/docs/theory/UseCases.md.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/8f2834ea8a657dacd79af87b24cd8e5135db646c

Im not sure if I can say anything more about it at this moment.

## 2. How does TrenchBoot work? (trenchboot architecture)

### Architecture Overview

*The full architecture overview can be found on
[the corresponding page](./theory/Architecture.md).*
![Overview](img/Architectural_Flow.png)
This diagram shows a general overview of the overall architecture of a
TrenchBoot system. In Bootstrap phase the system environment is measured and
that information stored in the dynamic PCRs, using whatever method is
appropriate for the given hardware. The intermediate phase analyzes the
information collected in the bootstrap phase to decide if it wants to continue
booting and pass control to the target OS, or if it wants to stop immediately.
If desired, additional evidence/measurements can be gathered in this phase. Once
 the integrity of the system is verified the TrenchBoot kernel passes off

>  control to the actual desired OS to initiate the runtime phase.

Reviewer 2: Remove spaces at the beginning of these lines.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/5b460bd264dc581af907ebe20e65c520e6f370ef#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R91-R92

### Bootstrap Phase - GRUB

> Typically, when attempting to establish a DRTM, TBoot is loaded in order to

Reviewer 2: We want TrenchBoot to be used instead of TBoot. Stating that `typically TBoot is loaded` doesn't help in that regard, nor is it relevant to how TrenchBoot works. Please drop this sentence and rephrase the next one accordingly.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/5b460bd264dc581af907ebe20e65c520e6f370ef#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R96-R101

carry out TXT on Intel, or SKINIT is used on AMD. TrenchBoot instead utilizes
its own system, which is a custom version of GRUB that has commands built in to
carry out a TrenchBoot "Secure Launch". This customized GRUB introduces the
`slaunch` and `slaunch_module` commands. A working example of the TrenchBoot
GRUB can be found at
[https://github.com/3mdeb/meta-trenchboot](
    https://github.com/3mdeb/meta-trenchboot).
The custom GRUB works with Linux and Multiboot2 boot protocols for both Intel
and AMD devices, but only in case of legacy boot, without support for UEFI

> (although it's available on a separate branch).

Reviewer 2: This is much more than just GRUB. If we really want to have this, it would fit better in "how to use TrenchBoot" section, but I think it should be removed instead. FAQ shouldn't need a build instructions and "where's the code" sections, these should be easy to find on the website.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/ac41b8036fee12265f4279ae44877302ae87e51b

### Intermediate Phase - Linux Kernel / U-ROOT (Initramfs)

The intermediate phase processes information collected in the bootstrap phase
and converts it into normalized data. This data is eventually used to determine
whether to boot into the target OS. The kernel and u-root initramfs
together make up Trenchboot's "Security Engine", which is an intermediate
mini-OS that serves to process data gathered by that bootstrap phase. During the
intermediate phase, the kernel and initramfs work together to measure block
devices, individual files, SMBUS/DMI information and provides unseal based,
external device and network based attestation. Trenchboot calls this
functionality secure launch (aka slaunch). The table in section 4 lists the
specific repo and branch to use when building the kernel or u-root.

## 3. How to use Trenchboot? (practical usage)

> ### Trenchboot GRUB TXT Config with Serial Communication

Reviewer 2: This seems out of place and incomplete.

Also, there is a link to user documentation [on the front page](https://github.com/TrenchBoot/TrenchBoot.github.io/blob/cleanup/docs/index.md?plain=1#L17), I was hoping that it would be enough to avoid questions like this one. Please remove whole `How to use Trenchboot?` part.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/1fe22a330c4672ea2ecaf9a90a9071edb1386b6a

Reviewer 2: Broken links, numbers after `#` were not changed on the list at the top.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/compare/25021ec141d2a88fe039a75c8743ba4dd733f2e3..3c9d6af92c149d463430036cddd65a3ba2e58968

*Note: ACM modules can be downloaded from [Intel's site][acm-download] (look for
"Production SINIT ACM Download").*

[acm-download]: https://www.intel.com/content/www/us/en/developer/articles/tool/intel-trusted-execution-technology.html

```text
menuentry 'TrenchBoot' {
        savedefault
        load_video
        set gfxpayload=keep
        insmod gzio
        insmod part_msdos
        insmod ext2
        set root='hd0,msdos1'
        multiboot2 /boot/slboot.gz loglvl=all serial=115200,8n1,0x3f8 logging=serial,memory
        module2 /boot/vmlinuz-trenchboot root=UUID=<your block uuid> ro intel_iommu=on loglevel=7 console=ttyS0,115200 nokaslr
        slaunch
        slaunch_module /boot/<your acm>.bin
}
```

> ### Testing the System

Reviewer 2: Drop this section. QEMU isn't able to emulate DRTM, and there are no instructions on how to obtain the images.

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/e3e8110fbfa16949a01ced6e56e04e5d501f62bf

can be found at
[https://github.com/3mdeb/meta-trenchboot](
    https://github.com/3mdeb/meta-trenchboot).
The custom GRUB works on Linux and Multiboot2 protocols for both Intel and AMD

> devices, but only in case of legacy boot, without support for UEFI.

Reviewer: Not sure about `with` but `on` makes it sound like GRUB is running in Linux.  I missed that since there are instructions for building GRUB for EFI from a different branch, its existence should probably me mentioned here too.
```suggestion
The custom GRUB works with Linux and Multiboot2 protocols for both Intel and AMD
devices, but only in case of legacy boot, without support for UEFI (available on a
separate branch).
```

Author: Notice this paragraph says that GRUB works on Linux protocols, not on Linux. I haven't heard about any Linux protocols though. Maybe it should say that it "works with Linux and on Multiboot2 protocol"?

Author: What do you think? https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/0151b8ba63954daca19d4b76590dae0684f6a449

Reviewer: Here "Linux protocol" means Linux boot protocol.  Maybe "boot" should be added to reduce confusion.

```suggestion
The custom GRUB works with Linux and Multiboot2 boot protocols for both Intel and AMD
devices, but only in case of legacy boot, without support for UEFI (although it's available on
a separate branch).
```

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/compare/0151b8ba63954daca19d4b76590dae0684f6a449..382f343dd9947180a69b693d3eb0720b69cfa754#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R94-R96

It therefore relies on the ability to know what correct is and to securely
verify measurement with expected correctness.

> ## 6. What do I need to incorporate TrenchBoot into my system?

Reviewer: This also makes no sense without an intermediate loader.

Author: Should this section be removed?

Reviewer: Let's keep this and other stuff which might not be relevant anymore until it's clear it won't be needed.  We can always remove it later.

an instance of the loader. After that, you configure your system boot to launch
the loader.

## 7. Where do I start if I want to help with contributions?

The [TrenchBoot Blueprints](./blueprints/index.md)
are how feature requests are collected for the project. Check if there is a
blueprint that is of your interest, if not, submit a blueprint via a pull
request for a feature you would like to see implemented.

## 8. Trenchboot Sources

| Name | Purpose | Functionality | Repo
|----------|-----|---|---
| Linux       | Kernel | The custom TrenchBoot linux kernel does include working code that allows u-root to function properly. This kernel also functions as a regular kernel in that when it is launched without SLBoot, the data required to begin the measured launch process is simply not exposed.   | link: [here](https://github.com/TrenchBoot/linux) working branch: linux-sl-5.5

> | U-root | initramfs/initrd  | By default launching u-root should show “u-root” in big bubble letters along with a shell. The u-root the command to launch into a MLE is sluinit which can be configured to be called automatically. | link: [here](https://github.com/TrenchBoot/u-root/) working branch: for-upstream

Reviewer: `U-root` looks abandoned (see other comments), https://github.com/TrenchBoot/xen/ and https://github.com/TrenchBoot/secure-kernel-loader/ are missing.

Author: Should U-root be removed from this table then?

Author: I am not sure what could be said about the `Functionality` of the projects you linked so I think I will refrain from pushing the table with empty fields?

Reviewer: I haven't actually read the info about Linux/GRUB before, but now I see that it's outdated and updating it can lead to duplication.  Better drop the table completely and direct the reader towards https://github.com/TrenchBoot and https://trenchboot.org/blueprints/.  And I've added links to repos in https://github.com/TrenchBoot/TrenchBoot.github.io/pull/42/files#diff-7453be7cde9748642eac972011134f410b4caf06c8fa323d4b48d03301868f2a.

| Grub | *experimental built-in* secure multiboot preboot module  |  The custom GRUB seems to be unfinished but soon aims to implement the slaunch and slaunch_module commands. | link: [here](https://github.com/TrenchBoot/grub) working branch: trenchboot_support_2.04

## 9. Various GitHub and further documentation links

- (AMD only project), contains TrenchBoot working custom GRUB and u-root

> with tpm2 tools 3mdeb, meta-trenchboot (2020, April 17) &ndash;

Reviewer: No longer AMD-only, but legacy-only as mentioned by another comment.  `3mdeb, meta-trenchboot` part looks unnecessary.

GRUB2 link below can be dropped, I'll have it in my PR along with other links to upstream activities for all projects.

Author: What GRUB2 link do you mean? I can't find it.
https://github.com/TrenchBoot/TrenchBoot.github.io/compare/77aaa7d4fca79f07460ba2bc9372c2e0cea0e01b..333691d2321d4a1a49e964dd41f811f4dda7daa5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R200-R201

Reviewer: My mistake, I meant the next bullet with Linux patches (`LKML Trenchboot doc`).

Author: https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41/commits/bacc6b5f14e77f60f13952fd96a65f6918b23b8a
